### DISCUSSION TITLE
ISO 19107

### DISCUSSION THREAD

**Discussion by AdrianKriger** [pyvista/pyvista#4604] [Q&A] [answered]
Thank you for this libray.

Do your polyhedra `pv.CellType.POLYHEDRON` create topologically correct solids that comply with the [ISO 19107 standard](https://www.iso.org/standard/66175.html) ~~ correct orientation of surfaces, no non manifold vertices, watertight volumes with no holes nor gaps, etc.) i.e.: parse through [val3dity](http://geovalidation.bk.tudelft.nl/val3dity/) with no errors?

**Comment by banesullivan** [comment, marked as answer]
I'm not familiar with that standard, but you can try experimenting with this example https://docs.pyvista.org/version/stable/examples/00-load/create-polyhedron.html

I've extracted the POLYHEDRON part to:

```py
import pyvista as pv

points = [
    [0.02, 0.0, 0.02],  # 0
    [0.02, 0.01, 0.02],  # 1
    [0.03, 0.01, 0.02],  # 2
    [0.035, 0.005, 0.02],  # 3
    [0.03, 0.0, 0.02],  # 4
    [0.02, 0.0, 0.03],  # 5
    [0.02, 0.01, 0.03],  # 6
    [0.03, 0.01, 0.03],  # 7
    [0.035, 0.005, 0.03],  # 8
    [0.03, 0.0, 0.03],  # 9
]

polyhedron_connectivity = [
    # NItems will go here
    7,  # number of faces
    5,  # number of points in face0
    0, # point index 0
    1, # point index 1
    2, # point index 2
    3, # point index 3
    4, # point index 4
    4, # # number of points in face1
    0, 1, 6, 5, # point index ...
    4, # n face2
    0, 4, 9, 5, # point index ...
    4, # n face3
    4, 9, 8, 3, # point index ...
    4, # n face4
    3, 8, 7, 2, # point index ...
    4, # n face5
    2, 7, 6, 1, # point index ...
    5, # n face6
    5, 6, 7, 8, 9, # point index ...
]

cells = [len(polyhedron_connectivity)] + polyhedron_connectivity
celltypes = [pv.CellType.POLYHEDRON]
grid = pv.UnstructuredGrid(cells, celltypes, points)

grid.plot(show_edges=True)
```

From what you describe, I think the VTK cell types adhers but you may want to post this on discourse.vtk.org as that community may have more insight.

Otherwise, it's tough to save and validate this object with that tool you shared as we currently only export UnstructuredGrid types to VTK-specific formats and the `meshio` extension complains about this 3D cell type. Many of the formats listed on that website, AFAIK, also don't support many 3D formats:

- OBJ: Wavefront .obj files can only contain triangle or quad cells.
- POLY: I'm assuming this is PLY which also is for tianglulated/poly meshes (2D geometries)
- OFF: only supports triangle cells.

**Reply by AdrianKriger** [reply]
Thank you

**Reply by MatthewFlamm** [reply]
It is worth pointing out that you can construct cells of type `POLYHEDRON` that might pass most or all of those criteria, but vtk and pyvista do not require that cells pass those criteria.  It is up to the user to ensure that cells are defined properly if they require it.

**Comment by sanek151** [comment]
> Thank you for this libray.
> 
> Do your polyhedra `pv.CellType.POLYHEDRON` create topologically correct solids that comply with the [ISO 19107 standard](https://www.iso.org/standard/66175.html) ~~ correct orientation of surfaces, no non manifold vertices, watertight volumes with no holes nor gaps, etc.) i.e.: parse through [val3dity](http://geovalidation.bk.tudelft.nl/val3dity/) with no errors?

123
